src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -1556,106 +1523,284 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
+            /// <returns>The string matching the hash, or null.</returns>
+            private static string CheckForMatches(uint[] runningHash, int upTo)
             {
-                yield return currentElement.ToString();
-            }
-
-            yield break;
-        }
+                var upToMax = runningHash.Length;
+                if (upTo == 0) upTo = upToMax;
+                else if (upTo > upToMax) upTo = upToMax;
 
-        // Regular string signatures that can be detected with just string comparison.
-        private static HashSet<string> s_signatures = new HashSet<string>(StringComparer.OrdinalIgnoreCase) {
-            // Calling Add-Type
-            "Add-Type", "DllImport",
+                for (var i = 0; i < upTo; i++)
+                {
+                    var result = LookupHash(runningHash[i]);
+                    if (result != null) return result;
+                }
 
-            // Doing dynamic assembly building / method indirection
-            "DefineDynamicAssembly", "DefineDynamicModule", "DefineType", "DefineConstructor", "CreateType",
-            "DefineLiteral", "DefineEnum", "DefineField", "ILGenerator", "Emit", "UnverifiableCodeAttribute",
-            "DefinePInvokeMethod", "GetTypes", "GetAssemblies", "Methods", "Properties",
+                return null;
+            }
 
-            // Suspicious methods / properties on "Type"
-            "GetConstructor", "GetConstructors", "GetDefaultMembers", "GetEvent", "GetEvents", "GetField",
-            "GetFields", "GetInterface", "GetInterfaceMap", "GetInterfaces", "GetMember", "GetMembers",
-            "GetMethod", "GetMethods", "GetNestedType", "GetNestedTypes", "GetProperties", "GetProperty",
-            "InvokeMember", "MakeArrayType", "MakeByRefType", "MakeGenericType", "MakePointerType",
-            "DeclaringMethod", "DeclaringType", "ReflectedType", "TypeHandle", "TypeInitializer",
-            "UnderlyingSystemType",
+            /// <summary>
+            /// Scan a string for suspicious content.
+            ///
+            /// This is based on the Rubin-Karp algorithm, but heavily
+            /// modified to support searching for multiple patterns at
+            /// the same time.
+            ///
+            /// The key difference from Rubin-Karp is that we don't undo
+            /// the hash of the first character as we shift along in the
+            /// input.
+            ///
+            /// Instead, we can rely on knowing we need the hashes for
+            /// shorter strings anyway, so we reuse their values in
+            /// computing the hash for the longer patterns. This lets us
+            /// use a much simpler hash as well - we can avoid the use of
+            /// mod.
+            /// </summary>
+            /// <returns>The string matching the hash, or null.</returns>
+            public static string Match(string text)
+            {
+                // The longest pattern is 29 characters.
```

Comment:
I assume the 29 character limit was chosen to accommodate the current set of suspicious patterns.  But what if a new pattern is added that is greater than 29 characters?  It seems the generator method (HashNewPattern) should check length and throw an error if greater than current longest pattern.

Reply:
Sure, I'll add a check.